pkg/config/validation/validation.go

```diff
@@ -1075,19 +1075,10 @@ func validateNamespaceSlashWildcardHostname(hostname string, isGateway bool) (er
 		errs = appendErrors(errs, fmt.Errorf("config namespace and dnsName in host entry cannot be empty"))
 	}
 
-	if !isGateway {
-		// namespace can be * or . or ~ or a valid DNS label in sidecars
-		if parts[0] != "*" && parts[0] != "." && parts[0] != "~" {
-			if !labels.IsDNS1123Label(parts[0]) {
-				errs = appendErrors(errs, fmt.Errorf("invalid namespace value %q in sidecar", parts[0]))
-			}
-		}
-	} else {
-		// namespace can be * or . or a valid DNS label in gateways
-		if parts[0] != "*" && parts[0] != "." {
-			if !labels.IsDNS1123Label(parts[0]) {
-				errs = appendErrors(errs, fmt.Errorf("invalid namespace value %q in gateway", parts[0]))
-			}
+	// namespace can be * or . or ~ or a valid DNS label in sidecars or gateways
+	if parts[0] != "*" && parts[0] != "." && parts[0] != "~" {
+		if !labels.IsDNS1123Label(parts[0]) {
+			errs = appendErrors(errs, fmt.Errorf("invalid namespace value %q", parts[0]))
```

Comment:
`~` namespace labels _are_ used by Istio Gateways when converting from Gateway API gateways and no matching routes can be found - I believe these generated resources were incorrectly failing validation before, see https://github.com/istio/istio/blob/e46a62df82104afdf4006775c344e690554cd7cc/pilot/pkg/config/kube/gateway/conversion.go#L2442-L2447

Reply:
The internal generate ones don't actually go through validation in real world -- only in the tests. 

Not sure how much we care about maintaining that property. Looks like in other parts we allowlist these "internal gateways" (see code like `gatewaySemantics := cfg.Annotations[constants.InternalRouteSemantics] == constants.RouteSemanticsGateway`)

Reply:
Rewrote this logic to plumb through a new `internal.istio.io/gateway-semantics` annotation, see latest commit.